contracts/FlowStakingCollection.cdc

```diff
@@ -0,0 +1,844 @@
+            // and locked vault capability
+            if let tokenHolderObj = tokenHolder {
+                self.tokenHolder = tokenHolder
+
+                // borrow the main token manager object from the locked account 
+                // to get access to the locked vault capability
+                let lockedTokenManager = tokenHolderObj.borrow()!.borrowTokenManager()
+                self.lockedVault = lockedTokenManager.vault
+            } else {
+                self.tokenHolder = nil
+                self.lockedVault = nil
+            }
+        }
+
+        /// Close all the stakes before destroying everything
+        /// This uses the closeStake method, so it will panic if there are still tokens staked in any of the objects
+        destroy() {
+            let nodeIDs = self.getNodeIDs()
+            let delegatorIDs = self.getDelegatorIDs()
+
+            for nodeID in nodeIDs {
+                self.closeStake(nodeID: nodeID, delegatorID: nil)
+            }
+
+            for delegatorID in delegatorIDs {
+                self.closeStake(nodeID: delegatorID.delegatorNodeID, delegatorID: delegatorID.delegatorID)
+            }
+
+            destroy self.nodeStakers
+            destroy self.nodeDelegators
+        }
+
+        /// Called when committing tokens for staking. Gets tokens from either or both vaults
+        /// Uses locked tokens first, then unlocked if any more are still needed
+        access(self) fun getTokens(amount: UFix64): @FungibleToken.Vault {
+
+            // If there is a locked account, use the locked vault first
+            if self.lockedVault != nil {
+
+                var lockedBalance: UFix64 = self.lockedVault!.borrow()!.balance - FlowStorageFees.minimumStorageReservation
+                var unlockedBalance: UFix64 = self.unlockedVault.borrow()!.balance - FlowStorageFees.minimumStorageReservation
+
+                assert(
+                    amount <= lockedBalance + unlockedBalance,
+                    message: "Insufficient total available Flow balance"
+                )
+
+                // If all the tokens can be removed from locked, withdraw and return them
+                if (amount <= lockedBalance) {
+                    self.lockedTokensUsed = self.lockedTokensUsed + amount
+
+                    return <-self.lockedVault!.borrow()!.withdraw(amount: amount)
+                
+                // If not all can be removed from locked, remove what can be, then remove the rest from unlocked
+                } else {
+
+                    // update locked tokens used record by adding the rest of the locked balance
+                    self.lockedTokensUsed = self.lockedTokensUsed + lockedBalance
+
+                    // Update the unlocked tokens used record by adding the amount requested
+                    // minus whatever was used from the locked tokens
+                    self.unlockedTokensUsed = self.unlockedTokensUsed + (amount - lockedBalance)
```

Comment:
Repeated calculation of `amount - lockedBalance` here and below, consider declaring a variable and using it in both places